Add Event Data Analytics/Aggregation Dialog

sydp opened the pull request:
**IMPORTANT: All Pull Requests should be connected to an issue, if you don't
have an issue, please start by creating an issue and link it to the PR.**

Please provide enough information so that others can review your pull request:

<!-- You can skip this if you're fixing a typo or adding a minor bug fix -->

+ What existing problem does this PR solve?
+ What new feature is being introduced with this PR?
+ Overview of changes to existing functions if required.

<!-- Example: This PR adds a function to do X, which adds the ability to do Y.
-->

**Checks**

- [ ] All tests succeed.
- [ ] Unit tests added.
- [ ] e2e tests added.
- [ ] Documentation updated.

**Closing issues**

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes
(if such).

jkppr commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+          field_query_string: this.eventValue
+        }
+      }).then((response) => {
+        this.stats = response.data.objects[0].field_summary.buckets[0]
```
During testing your PR I requarly got the following error on the console when loading aggregations for various fields:
```
Uncaught (in promise) TypeError: Cannot read properties of null (reading 'type')
```
My theory on this problem: It has to do with some race condition where `stats` is not yet defined by the mounted method call and the graphs already trying to call the computed methods. 

On way to prevent this race condition that I can think of:
1. Add a new variable `statsReady: false`to the data above.
2. Set this variable to true when the promise here is fulfilled.
3. Add the `:loading="!statsReady"` option to the `v-card` elements to display a loading animation when the data is not ready yet. 
4. Add `v-if="statsReady"` to the `v-card-text` elements that contain the apexchart or to the apexchart itself to only render it when the data is available.

At least in my tests, this did prevent all errors and also shows a nice loading animation with aggregations on larger data sets.

sydp replied:
Thanks I have implemented the proposed solution and it seems to work nicely.

jkppr reviewed (commented):
Great work on the aggregations! I'm already a fan and look forward to use them with the next investigation.
I have added some UX/UI feedback as also some bug fixes, please take a look.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+                Value as a percentage of {{ this.eventKey }} events
+              </v-card-title>
+              <v-card-text>
+                <!-- {{  this.valueEventCount }} -->
```
Done, thx!

sydp commented on timesketch/frontend-ng/src/components/Explore/EventDetail.vue:
```diff
@@ -35,10 +35,18 @@ limitations under the License.
                       <v-icon small>mdi-content-copy</v-icon>
                     </v-btn>
-
+                    <v-btn
```
Done

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+      <v-container fluid>
+        <v-row justify="center">
+          <v-col>
+            <v-card>
```
Yes agree, I set the v-card height as suggested.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+      return this.distributionIntervals[this.selectedDistributionIntervalIndex]
+    },
+    intervalChartOptions() {
+      if (this.stats === undefined) return []
```
Done

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+<template>
```
Done, i have added the "X".  I will need to implement the close logic.

jkppr replied:
For the logic, take a look at the implementation of the `FormatXMLString.vue` dialog.

Basically you add a `@cancel="aggregatorDialog = false"` to the `v-dialog` and trigger it via a method in the `AggregateDialog.vue` file by executing `this.$emit('cancel')`.

sydp replied:
Thanks for the tip, have got this working now!

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+            <v-card>
+              <v-card-title>
+                Value event distribution by {{ this.distributionIntervals[this.selectedDistributionIntervalIndex] }}
+              </v-card-title>
```
For the percentage donut chart, I updated the card title to make it a bit clearer and adding the raw values to the labels.

sydp replied:
And I added the v-card-subtitle of the value and updated the titles.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+          <v-col align="center">
+            <v-card>
+              <v-card-title>
+                Value as a percentage of {{ this.eventKey }} events
```
Done.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+    <v-card-text>
+      <v-toolbar flat>
+        <v-toolbar-title>
+          <div class="font-weight-bold text-h5">Event Data Analytics</div>
```
Done. 

The toolbar was a used in a previous iteration of the component (for buttons to control a larger histogram/heatmap of aggregate event counts) but now that these are no longer there, the toolbar is no longer needed and the suggested changes look much better.  Cheers.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,399 @@
+-->
+<template>
+  <v-card class="mx-auto">
+    <v-card-text>
```
I used the v-card-title and rearranged the elements.

sydp commented:
@jkppr - thanks for the initial review and feedback.  I've made some changes as suggested, please take a look when you can

jkppr commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+      <v-container fluid>
+        <v-row justify="center">
+          <v-col>
+            <v-card height="450px" :loading="!statsReady">
```
I took a stab at the general statistics card, since it looks a bit lost visually compared to the other elements and I have a hard time to read/understand the data. 

What do you think about this approach:
![general_stats](https://user-images.githubusercontent.com/99879757/226875401-ed7b1edb-0ad0-41ea-a751-376d11632f5b.png)

If you like it, here is the code I used. Feel free to adjust it to your preferences.
```
<v-col>
  <v-card height="145px" :loading="!statsReady">
    <v-card-title>
      Sketch statistics
    </v-card-title>
    <v-simple-table dense v-if="statsReady" class="px-2 mt-n4">
      <tbody>
        <tr>
          <td width="360px">Total number of events</td>
          <td><strong>{{ this.docCount }}</strong></td>
        </tr>
        <tr>
          <td width="360px">First event</td>
          <td><strong>{{ this.fieldDateTimeMinimum }}</strong></td>
        </tr>
        <tr>
          <td width="360px">Last event</td>
          <td><strong>{{ this.fieldDateTimeMaximum }}</strong></td>
        </tr>
      </tbody>
    </v-simple-table>
  </v-card>
  <v-card height="145px" :loading="!statsReady" class="mt-1">
    <v-card-title>
      Field statistics
    </v-card-title>
    <v-simple-table dense v-if="statsReady" class="px-2 mt-n4">
      <tbody>
        <tr>
          <td width="360px">Field name</td>
          <td><span style="font-family: monospace">{{ this.eventKey }}</span></td>
        </tr>
        <tr>
          <td width="360px">Total number of events</td>
          <td><strong>{{ this.fieldValueCount }}</strong></td>
        </tr>
        <tr>
          <td width="360px">Number of unique values</td>
          <td><strong>{{ this.fieldCardinality }}</strong></td>
        </tr>
      </tbody>
    </v-simple-table>
  </v-card>
  <v-card height="180px" :loading="!statsReady" class="mt-1">
    <v-card-title>
      Value statistics
    </v-card-title>
    <v-simple-table dense v-if="statsReady" class="px-2 mt-n4">
      <tbody>
        <tr>
          <td width="360px">Field value</td>
          <td><span style="font-family: monospace">{{ this.eventValue }}</span></td>
        </tr>
        <tr>
          <td width="360px">Total number of events</td>
          <td><strong>{{ this.valueEventCount }}</strong></td>
        </tr>
        <tr>
          <td width="360px">First event</td>
          <td><strong>{{ this.valueDateTimeMinimum }}</strong></td>
        </tr>
        <tr>
          <td width="360px">Last event</td>
          <td><strong>{{ this.valueDateTimeMaximum }}</strong></td>
        </tr>
      </tbody>
    </v-simple-table>
  </v-card>
</v-col>
```

sydp replied:
thanks, i was having a think today and wasn't quite sure if these values needed an additional visualisation or formatting but agree this looks better.

berggren replied:
Sketch statistics is a bit malplaced IMO. I would not expect to get that information when I click on a key/value.

jkppr replied:
I would argue that the general information on how many events a sketch has and which time window the sketch covers provides a good reference point for the analyst to set a context on the statistic values provided for field and value.

sydp replied:
I made the suggested changes with some tweaks on column width.  

The idea for having the sketch statistics is as described by @jkppr - to get a reference of the selected value time frame relative to the entire sketch.  Another option would be to use those values in another visualisation or two?

Let me know how you would like to make additional changes.

jkppr replied:
Since I don't see this as blocking, I would keep the sketch statistics for now as context information. Let's collect some user feedback when this PR is in and then revisit it in the second iteration.

jkppr replied:
@sydp small UI nit: At least on my system the Sketch statistics and Field statistics v-card are 1px too small in height resulting in the table overlaying the bottom border line.

![general_stats](https://user-images.githubusercontent.com/99879757/227532218-56de3473-a99a-412f-ac70-c309783f7d01.png)

sydp replied:
I made some adjustments to reveal the border lines

jkppr reviewed (changes requested):
Some final visual polishing, comments & suggestions, please take a look. I think after this round we should be ready to merge.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+            <v-col align="center">
+            <v-card height="500px" :loading="!dataReady">
+              <v-card-title>
+                Surrounding events
```
"Surrounding events" is meant to represent the aggregate count of events +/- the selected time interval (year, month, etc) around the specific event.    

As an example, if a user selects the aggregate dialog from the event detail view for an "event_id: 4624" at time 2020-02-02, then this visualisation is meant to show this event as the "center" of the graph and then other "event_id:4624" event buckets around that 2020-02-02 time +/- the selected time interval.

In hindsight, I excluded buckets with zero events (i.e. you only see bars for time intervals where there are matching events), you lose some contextual information about where events are relative to the current event.  I will make the adjustment to include zero buckets which hopefully should make it more understandable.

berggren replied:
Cool idea. This is essentially the `context query` look-around feature that we are still to port/redesign.

I was also struggling to understand it. If we keep it, I think we need to add a descriptive text to it. 

+-5 years is a long span to default to. Can we make that say 7 days as a default?

sydp replied:
and now including empty buckets.

sydp replied:
Done.

sydp commented:
> Some final visual polishing, comments & suggestions, please take a look. I think after this round we should be ready to merge.

cheers, will take a look at addressing these changes tomorrow.

berggren commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+    </v-card-subtitle>
+    <v-card-text>
+      <v-container fluid>
+        <v-row justify="center">
```
Good suggestion. Also, I would prefer the dialog to fit for a laptop screen if possible :) When I test this with a bit lower resolution I have to scroll the see al the data.

berggren commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+                  Top {{ Math.min(10, this.commonValues.length) }} &nbsp;<span style="font-family: monospace">{{ eventKey }}</span>&nbsp; values (out of {{ this.fieldCardinality }})
+                </v-card-title>
+                <v-card-text>
+                  <v-data-table
```
10 is the default, we can return more. What about returning all and keep the pagination?

jkppr replied:
Both works for me, as long as we don't call it Top 10 anymore if we go with all entries.

sydp replied:
Done (added the pagination footer check).

For the top/most common values, the results will be limited to 10 since that is an open search parameter for this aggregation (which defaults to 10).  Because of this, I would opt to leave it to 10 (or similar number) for a few reasons:
* returning "all" will introduce additional logic to split between "common" and "rare"
* for certain numeric and text fields, the number of values could be very high and it wouldn't be useful to navigate these values "10 at a time".  
 
For the rare values, the results are limited by document count (i.e. number of events).  This is currently set to maximum of 5 so in some cases the number of results could vary significantly.

jkppr replied:
Given top 10 is the default and should cover most needs for an analyst I will resolve this comment. We can revisit the default value or make them even choosable by the analyst in the next iteration.

berggren commented on timesketch/frontend-ng/src/components/Explore/EventDetail.vue:
```diff
@@ -228,18 +237,33 @@ limitations under the License.
+        :eventValue="eventValue"
+        :eventTimestamp="eventTimestamp"
+        :eventTimestampDesc="eventTimestampDesc"
+        :sketchId="this.sketch.id"
```
You don't need this. You can use a sketch computed property in the dialog component instead.

sydp replied:
Done

berggren commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+-->
+<template>
+  <v-card class="mx-auto">
+    <v-card-title>
+      <span class="headline">Event Data Analytics</span>
+      <v-spacer></v-spacer>
+      <v-btn icon @click="clearAndCancel">
+        <v-icon>mdi-close</v-icon>
+      </v-btn>
+    </v-card-title>
```
In other dialogs (see the graph bottom sheet for example) we use a v-toolbar. This will make sure that the toolbar is visible when scrolling etc.

```
        <v-toolbar dense flat>
          <strong>Title</strong>
          <v-spacer></v-spacer>
          <v-btn icon @click="showTimelineView = false">
            <v-icon>mdi-close</v-icon>
          </v-btn>
        </v-toolbar>
        ```

sydp replied:
I found a solution to ensure the toolbar is visible when scrolling.

jkppr replied:
Cool solution with the toolbar that stays on-top when scrolling 👍  
However, now the field & value information are missing that have been in the subtitle before. I personally liked it, since it gave an immediate overview what field and value I'm looking at. Maybe you have a good idea on how to bring it back into the toolbar? What do you think?

berggren commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+      <v-container fluid>
+        <v-row justify="center">
+          <v-col>
+            <v-card height="450px" :loading="!statsReady">
```
We don't use the elevation (shadow) on cards in the UI. add the `outlined` attribute to switch to the flat version.

sydp replied:
Done

berggren reviewed (dismissed):
First round, PTAL

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+                  Rare &nbsp;<span style="font-family: monospace">{{ eventKey }}</span>&nbsp; values (max count of 5)
+                </v-card-title>
+                <v-card-text>
+                  <v-data-table
```
Yes, that's correct.  Done.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+          <v-col align="center">
+            <v-card height="450px" :loading="!statsReady">
+              <v-card-title>
+                Count of &nbsp;<span style="font-family: monospace">{{ eventValue }}</span>&nbsp; as a percentage of &nbsp;<span style="font-family: monospace">{{ eventKey }}</span>&nbsp; events
```
Done

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,537 @@
+              </v-card-title>
+              <v-card-text v-if="statsReady">
+                <apexchart
+                  height="250px"
```
Done.

sydp commented:
> > Some final visual polishing, comments & suggestions, please take a look. I think after this round we should be ready to merge.
> 
> cheers, will take a look at addressing these comments tomorrow.

I made some changes as suggested/per review.  PTAL when you get a chance, thanks! :)

jkppr commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,622 @@
+                  <template v-slot:activator="{ on }">
+                    <v-icon color="primary" v-on="on">mdi-information-outline</v-icon>
+                  </template>
+                  <span>Shows the distribution of &nbsp;<span style="font-family: monospace">{{ eventValue }}</span>&nbsp; 
```
```suggestion
                  <span>Shows the distribution of <span style="font-family: monospace">{{ this.truncateValue(eventValue) }}</span>
```
* Remove the `&nbsp;` and adds the `truncateValue()` function.

jkppr commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,622 @@
+          enabled: true
+        },
+        labels: [
+          this.eventValue + ' (' + this.valueEventCount + ')',
```
```suggestion
          this.truncateValue(this.eventValue)  + ' (' + this.valueEventCount + ')',
```
* The `truncateValue()` function can also be applied to labels to make them more readable.

jkppr reviewed (changes requested):
Thanks for all the changes and work so far. 

I have some more suggestions from the visual point of view. Please have a look.

berggren reviewed (commented):
Overall it looks good. I let @jkppr finish the review and merge when ready.

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,622 @@
+                </tbody>
+              </v-simple-table>
+            </v-card>
+            <v-card outlined height="147px" :loading="!statsReady" class="mt-1">
```
I tested and added this change locally

sydp commented on timesketch/frontend-ng/src/components/Explore/AggregateDialog.vue:
```diff
@@ -0,0 +1,622 @@
+                </v-card-text>
+              </v-card>
+            </v-col>
+            <v-col align="center">
```
I removed this class from the other v-col to align the cards on my dev setup.  Let me know if this is still required.

jkppr reviewed (approved):
Awesome work 🎉  
LGTM, I'll merge it today.
